src/build/build_step.go

```diff
@@ -1131,6 +1136,65 @@ func fetchOneRemoteFile(state *core.BuildState, target *core.BuildTarget, url st
 	return nil
 }
 
+// setHeaders sets up all the headers we should send on remote_file() requests, including User-Agent and any user
+// defined ones.
+func setHeaders(req *http.Request, target *core.BuildTarget, env core.BuildEnv) error {
+	req.Header.Set("User-Agent", fmt.Sprintf("please.build/%s", core.PleaseVersion))
+
+	param := func(str string) (string, string) {
+		if !strings.HasPrefix(str, "remote_file:") {
+			return "", ""
+		}
+		str = strings.TrimPrefix(str, "remote_file:")
+		i := strings.IndexRune(str, ':')
+		return str[:i], str[(i + 1):]
+	}
+
+	userName := ""
+	password := ""
+	for _, l := range target.Labels {
+		param, value := param(l)
+		switch param {
+		case "":
+			continue
+		case "header":
+			k, v := header(value)
+			log.Warningf(v)
+			v = os.Expand(v, env.ReplaceEnvironment)
+			log.Warningf("setting header: %v: %v", k, v)
+			req.Header.Set(k, v)
+		case "secret_header":
+			k, v := header(value)
+			b, err := os.ReadFile(fs.ExpandHomePath(v))
+			if err != nil {
+				return fmt.Errorf("failed to read secret file: %v", err)
+			}
+
+			req.Header.Set(k, string(b))
+		case "username":
+			userName = value
+		case "password_file":
+			p, err := os.ReadFile(fs.ExpandHomePath(value))
+			if err != nil {
+				return fmt.Errorf("failed to read password file: %v", err)
+			}
+			password = string(p)
+		default:
+			return fmt.Errorf("unknown remote file label: %v", l)
+		}
+	}
+
+	if userName != "" {
```

Comment:
Might be worth checking `password` here too?

Reply:
I don't think it's necessarily wrong to have an empty password.